common/src/misc/logInDelegate/apiLogin.delegate.ts

```diff
@@ -0,0 +1,72 @@
+import { LogInDelegate } from "./logIn.delegate";
+
+import { TokenRequestTwoFactor } from "../../models/request/identityToken/tokenRequest";
+
+import { ApiService } from "../../abstractions/api.service";
+import { AppIdService } from "../../abstractions/appId.service";
+import { CryptoService } from "../../abstractions/crypto.service";
+import { EnvironmentService } from "../../abstractions/environment.service";
+import { LogService } from "../../abstractions/log.service";
+import { MessagingService } from "../../abstractions/messaging.service";
+import { PlatformUtilsService } from "../../abstractions/platformUtils.service";
+import { StateService } from "../../abstractions/state.service";
+import { TokenService } from "../../abstractions/token.service";
+import { TwoFactorService } from "../../abstractions/twoFactor.service";
+
+import { KeyConnectorService } from "../../abstractions/keyConnector.service";
+import { ApiTokenRequest } from "../../models/request/identityToken/apiTokenRequest";
+import { IdentityTokenResponse } from "../../models/response/identityTokenResponse";
+
+export class ApiLogInDelegate extends LogInDelegate {
+  tokenRequest: ApiTokenRequest;
+
+  constructor(
+    cryptoService: CryptoService,
+    apiService: ApiService,
+    tokenService: TokenService,
+    appIdService: AppIdService,
+    platformUtilsService: PlatformUtilsService,
+    messagingService: MessagingService,
+    logService: LogService,
+    stateService: StateService,
+    setCryptoKeys = true,
+    twoFactorService: TwoFactorService,
+    private environmentService: EnvironmentService,
+    private keyConnectorService: KeyConnectorService
+  ) {
+    super(
+      cryptoService,
+      apiService,
+      tokenService,
+      appIdService,
+      platformUtilsService,
+      messagingService,
+      logService,
+      stateService,
+      twoFactorService,
+      setCryptoKeys
+    );
+  }
+
+  async init(clientId: string, clientSecret: string, twoFactor?: TokenRequestTwoFactor) {
```

Comment:
A constructor can't be `async`, but what I've done instead is create an async static factory method. This creates the delegate object, initializes it, and returns it ready to be used. The constructor of each delegate is now `private` to prevent it from being created any other way.

Reply:
yeah I missed that it was due to `async`, but I still think that's the proper way to solve this